src/Draco.Compiler/Internal/Syntax/ParseTreeFormatter.cs

```diff
@@ -0,0 +1,185 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Draco.Compiler.Api.Syntax;
+using Token = Draco.Compiler.Internal.Syntax.ParseTree.Token;
+
+namespace Draco.Compiler.Internal.Syntax;
+
+internal sealed record class ParseTreeFormatterSettings(string Indentation);
+
+internal class ParseTreeFormatter : ParseTreeTransformerBase
+{
+    private TokenType? lastToken;
+    private TokenType? nextToken;
+    private IEnumerator<Token>? tokens;
+    private int indentCount = 0;
+    private ParseTreeFormatterSettings settings;
+    private string Indentation
+    {
+        get
+        {
+            var result = new StringBuilder();
+            for (int i = 0; i < this.indentCount; i++, result.Append(this.settings.Indentation)) ;
+            return result.ToString();
+        }
+    }
```

Comment:
I'd move down `indentCount` to be closer to `Indentation`, now the order is a bit messed up